### PR TITLE
add keystone testing

### PR DESCRIPTION
**Pull request by gtmanfred**

### CODE REVIEW

File: apache/init.sls [hunk 1]
```diff
@@ -15,7 +15,7 @@
     - require:
       - pkg: {{ apache }}
 
-{%- if grains['os_family'] == 'RedHat' %}
+{%- if grains['os_family'] == 'RedHat' and not grains['osmajorrelease'].startswith('7') %}
```
Comment (jtand) [inline apache/init.sls]: We don't need rhel/cent7 for this?

**Comment by cachedout** [comment]
Did this get tested? It is totally broken on Ubuntu 16 at least:

```
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('include', ['.setup_db'])])
[PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/keystone/init.sls' using 'yaml' renderer: 0.375347852707
[DEBUG   ] In saltenv 'base', looking at rel_path 'keystone/setup_db.sls' to resolve 'salt://keystone/setup_db.sls'
[DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/master/files/base/keystone/setup_db.sls' to resolve 'salt://keystone/setup_db.sls'
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://keystone/setup_db.sls'
[DEBUG   ] No dest file found
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/setup_db.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/setup_db.sls', mode set to 0644
[DEBUG   ] compile template: /var/cache/salt/master/files/base/keystone/setup_db.sls
[DEBUG   ] Jinja search path: ['/var/tmp/.root_9a762c_salt/running_data/var/cache/salt/minion/files/base']
[PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/keystone/setup_db.sls' using 'jinja' renderer: 0.00188088417053
[DEBUG   ] Rendered data from file: /var/cache/salt/master/files/base/keystone/setup_db.sls:
include:
  - .config

setup keystone db:
  file.directory:
    - name: /var/lib/keystone
    - user: keystone
    - group: keystone
    - dir_mode: 0700

  cmd.run:
    - name: keystone-manage db_sync
    - runas: keystone
    - unless: false

  grains.present:
    - name: setup_keystone_db
    - value: true
    - require:
      - cmd: setup keystone db

setup keystone fernet:
  file.directory:
    - name: /etc/keystone/fernet-keys/
    - user: keystone
    - group: keystone
    - dir_mode: 0700

  cmd.run:
    - name: keystone-manage fernet_setup --keystone-user keystone --keystone-group keystone
    - runas: keystone
    - unless: false

  grains.present:
    - name: setup_keystone_fernet
    - value: true
    - require:
      - cmd: setup keystone fernet

[DEBUG   ] Results of YAML rendering: 
OrderedDict([('include', ['.config']), ('setup keystone db', OrderedDict([('file.directory', [OrderedDict([('name', '/var/lib/keystone')]), OrderedDict([('user', 'keystone')]), OrderedDict([('group', 'keystone')]), OrderedDict([('dir_mode', 700)])]), ('cmd.run', [OrderedDict([('name', 'keystone-manage db_sync')]), OrderedDict([('runas', 'keystone'zjenkins-ubuntu16-05631e:
    - Rendering SLS 'base:keystone.package' failed: Jinja error: keystone/defaults.yaml
      Traceback (most recent call last):
        File "/home/mp/devel/salt/salt/utils/templates.py", line 368, in render_jinja_tmpl
          output = template.render(**decoded_context)
        File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
          return self.environment.handle_exception(exc_info, True)
        File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
          reraise(exc_type, exc_value, tb)
        File "<template>", line 1, in top-level template code
        File "/home/mp/devel/salt/salt/utils/jinja.py", line 135, in get_source
          raise TemplateNotFound(template)
      TemplateNotFound: keystone/defaults.yaml
      
      ; line 1
      
      ---
      {% from "keystone/defaults.yaml" import rawmap with context %}    <======================
      {%- set keystone = salt['grains.filter_by'](rawmap, grain='os') %}
      
      include:
        - apache
        - openstack.repo
      [...]
      ---
)]), OrderedDict([('unless', False)])]), ('grains.present', [OrderedDict([('name', 'setup_keystone_db')]), OrderedDict([('value', True)]), OrderedDict([('require', [OrderedDict([('cmd', 'setup keystone db')])])])])])), ('setup keystone fernet', OrderedDict([('file.directory', [OrderedDict([('name', '/etc/keystone/fernet-keys/')]), OrderedDict([('user', 'keystone')]), OrderedDict([('group', 'keystone')]), OrderedDict([('dir_mode', 700)])]), ('cmd.run', [OrderedDict([('name', 'keystone-manage fernet_setup --keystone-user keystone --keystone-group keystone')]), OrderedDict([('runas', 'keystone')]), OrderedDict([('unless', False)])]), ('grains.present', [OrderedDict([('name', 'setup_keystone_fernet')]), OrderedDict([('value', True)]), OrderedDict([('require', [OrderedDict([('cmd', 'setup keystone fernet')])])])])]))])
[PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/keystone/setup_db.sls' using 'yaml' renderer: 0.45256114006
[DEBUG   ] In saltenv 'base', looking at rel_path 'keystone/config.sls' to resolve 'salt://keystone/config.sls'
[DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/master/files/base/keystone/config.sls' to resolve 'salt://keystone/config.sls'
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://keystone/config.sls'
[DEBUG   ] No dest file found
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/config.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/config.sls', mode set to 0644
[DEBUG   ] compile template: /var/cache/salt/master/files/base/keystone/config.sls
[DEBUG   ] Jinja search path: ['/var/tmp/.root_9a762c_salt/running_data/var/cache/salt/minion/files/base']
[PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/keystone/config.sls' using 'jinja' renderer: 0.000998973846436
[DEBUG   ] Rendered data from file: /var/cache/salt/master/files/base/keystone/config.sls:
include:
  - .package

keystone config:
  file.managed:
    - name: /etc/keystone/keystone.conf
    - makedirs: True
    - template: jinja
    - source: salt://keystone/files/keystone.conf.jinja

  grains.present:
    - name: keystone.endpoint
    - value: http://127.0.0.1:35357/v2.0

[DEBUG   ] Results of YAML rendering: 
OrderedDict([('include', ['.package']), ('keystone config', OrderedDict([('file.managed', [OrderedDict([('name', '/etc/keystone/keystone.conf')]), OrderedDict([('makedirs', True)]), OrderedDict([('template', 'jinja')]), OrderedDict([('source', 'salt://keystone/files/keystone.conf.jinja')])]), ('grains.present', [OrderedDict([('name', 'keystone.endpoint')]), OrderedDict([('value', 'http://127.0.0.1:35357/v2.0')])])]))])
[PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/keystone/config.sls' using 'yaml' renderer: 0.00218510627747
[DEBUG   ] In saltenv 'base', looking at rel_path 'keystone/package.sls' to resolve 'salt://keystone/package.sls'
[DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/master/files/base/keystone/package.sls' to resolve 'salt://keystone/package.sls'
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://keystone/package.sls'
[DEBUG   ] No dest file found
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/package.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'keystone/package.sls', mode set to 0644
[DEBUG   ] compile template: /var/cache/salt/master/files/base/keystone/package.sls
[DEBUG   ] Jinja search path: ['/var/tmp/.root_9a762c_salt/running_data/var/cache/salt/minion/files/base']
[DEBUG   ] LazyLoaded roots.envs
[DEBUG   ] Could not LazyLoad roots.init: 'roots.init' is not available.
[DEBUG   ] Could not find file 'salt://keystone/defaults.yaml' in saltenv 'base'
[DEBUG   ] Jinja Error
[DEBUG   ] Exception: keystone/defaults.yaml
[DEBUG   ] Out: 
[DEBUG   ] Line: 1
[DEBUG   ] TmplStr: {% from "keystone/defaults.yaml" import rawmap with context %}
{%- set keystone = salt['grains.filter_by'](rawmap, grain='os') %}

include:
  - apache
  - openstack.repo

keystone packages:
  pkg.latest:
    - force_yes: True
    - pkgs:
      - {{ keystone.package }}
      - {{ keystone.wsgi }}
      - {{ keystone.apache }}
      - python-keystoneclient

  service.dead:
    - name: {{keystone.service}}

  file.managed:
    - name: {{keystone.apache_dir}}/keystone.conf
    - contents: |
        ServerName controller
        Listen 5000
        Listen 35357

        <VirtualHost *:5000>
            WSGIDaemonProcess keystone-public processes=1 threads=1 user=keystone group=keystone display-name=%{GROUP}
            WSGIProcessGroup keystone-public
            WSGIScriptAlias / /usr/bin/keystone-wsgi-public
            WSGIApplicationGroup %{GLOBAL}
            WSGIPassAuthorization On
            ErrorLogFormat "%{cu}t %M"
            ErrorLog /var/log/{{keystone.apache}}/keystone-error.log
            CustomLog /var/log/{{keystone.apache}}/keystone-access.log combined

            <Directory /usr/bin>
                Require all granted
            </Directory>
        </VirtualHost>

        <VirtualHost *:35357>
            WSGIDaemonProcess keystone-admin processes=1 threads=1 user=keystone group=keystone display-name=%{GROUP}
            WSGIProcessGroup keystone-admin
            WSGIScriptAlias / /usr/bin/keystone-wsgi-admin
            WSGIApplicationGroup %{GLOBAL}
            WSGIPassAuthorization On
            ErrorLogFormat "%{cu}t %M"
            ErrorLog /var/log/{{keystone.apache}}/keystone-error.log
            CustomLog /var/log/{{keystone.apache}}/keystone-access.log combined

            <Directory /usr/bin>
                Require all granted
            </Directory>
        </VirtualHost>
    - listen_in:
      - service: {{keystone.apache}}

{%- if grains['os'] == 'Ubuntu' %}
  apache_module.enabled:
    - name: wsgi
    - listen_in:
      - service: {{keystone.apache}}
{%- endif %}

Traceback (most recent call last):
  File "/home/mp/devel/salt/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/home/mp/devel/salt/salt/utils/jinja.py", line 135, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: keystone/defaults.yaml

[DEBUG   ] TraceStr: Traceback (most recent call last):
  File "/home/mp/devel/salt/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/home/mp/devel/salt/salt/utils/jinja.py", line 135, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: keystone/defaults.yaml

[ERROR   ] Rendering exception occurred: Jinja error: keystone/defaults.yaml
Traceback (most recent call last):
  File "/home/mp/devel/salt/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/home/mp/devel/salt/salt/utils/jinja.py", line 135, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: keystone/defaults.yaml

; line 1

---
{% from "keystone/defaults.yaml" import rawmap with context %}    <======================
{%- set keystone = salt['grains.filter_by'](rawmap, grain='os') %}

include:
  - apache
  - openstack.repo
[...]
---
[CRITICAL] Rendering SLS 'base:keystone.package' failed: Jinja error: keystone/defaults.yaml
Traceback (most recent call last):
  File "/home/mp/devel/salt/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/home/mp/devel/salt/salt/utils/jinja.py", line 135, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: keystone/defaults.yaml

; line 1

---
{% from "keystone/defaults.yaml" import rawmap with context %}    <======================
{%- set keystone = salt['grains.filter_by'](rawmap, grain='os') %}

include:
  - apache
  - openstack.repo
```

**Comment by gtmanfred** [comment]
I will check it this morning, but I tested it on 14.04 16.04 , red hat 7
and Centos 7, the only 4 distros that mitaka is supported on

**Comment by cachedout** [comment]
OK. It's blowing up for me when trying to deploy using the `--interactive` switch to salt-jenkins. Could be related to that since it's obviously deploying correctly on Jenkins itself.

**Pull request merged**